### PR TITLE
[SPARK-28188] Materialize Dataframe API

### PR DESCRIPTION
**Pull request by vgankidi**
## What changes were proposed in this pull request?
Add materialize API to dataframes. Materialize is a Spark action. It is a way to let Spark explicitly know that the dataframe has already been computed. Once a dataframe is materialized, Spark skips all stages prior to the materialize when the dataframe is reused later on. 
Please refer to [SPARK-28188](https://issues.apache.org/jira/browse/SPARK-28188) for the rationale behind adding the materialize API 

## How was this patch tested?
Tested manually

### CODE REVIEW

**Comment by srowen** [comment]
I don't think we should add this. It's already very common to `.count()` or `.mapPartitions` with a no-op to do this. I do think there are use cases for proactively materializing, but, it's overused too.

**Comment by danielcweeks** [comment]
@srowen I think the point here is to provide an explicit way of materializing data so that it can be reused in later stages as opposed to relying on side-effects of operations that produce the same result.

This is a really common issue, so having a clear expression of the intent really helps to disambiguate the logic.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -209,6 +209,11 @@ abstract class RDD[T: ClassTag](
    */
   def cache(): this.type = persist()
 
+  def materialize(): RDD[T] = withScope{
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Please fix the PR title. This is RDD API, not a `Dataframe` API.

**Comment by dongjoon-hyun** [comment]
I agree with @srowen , but since this is a legitimate community request, ping @mateiz , @rxin , @gatorsmile , @cloud-fan , too.

**Comment by rxin** [comment]
Did I miss something? How does "runJob" materialize the query plan / rdd?

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 1 above)
Comment (rdblue) [reply core/src/main/scala/org/apache/spark/rdd/RDD.scala]: This PR also includes `Dataset.materialize`, which calls this RDD version.

**Comment by rdblue** [comment]
@rxin, this runs the query up to the point where `materialize` is called. The underlying RDD can then pick up from the last shuffle the next time it is used. This works better than caching in most cases when using dynamic allocation because executors are not sitting idle, but work can be resumed and shared across queries. We could rename the method if that would be more clear.

@srowen, I've seen this suggested on the dev list a few times and I think it is a good idea to add it. There is not guarantee that `count` does the same thing -- it could be optimized -- and it is a little tricky to  get this to work with the dataset API. This version creates a new DataFrame from the underlying RDD so  that the work is reused from the last shuffle, instead of allowing the planner to re-optimize with later changes (usually projections) and discard the intermediate result. We have found this really useful for better control over the planner, as well as to cache data using the shuffle system.

**Comment by rxin** [comment]
Got it. But the name is plain wrong because the function doesn't materialize the DataFrame. As a matter of fact, if you run this "materialize" on a query plan without a shuffle this becomes a job that's completely useless that waste a lot of resources.

I've seen a different use case myself: I want to measure the execution time of a query plan, without materializing the data or invoking any I/O, or any overhead. What I ended up doing was implementing a simple data sink that doesn't write anything.

Looks like that can be used here as well?

**Comment by cloud-fan** [comment]
+1 for the no-op sink, I think it should be the same as the no-op `runJob` here.

**Comment by vgankidi** [comment]
@rxin Yes, materializing on a query plan without a shuffle just wastes resources. We usually recommend  repartitioning the dataframe before invoking materialize to add a shuffle. I think this can be used for your use case of measuring execution time of a query plan. What name would you suggest for this api?

**Comment by rxin** [comment]
Something that's like write to a no-op sink ...

**Comment by vgankidi** [comment]
Updated the PR with the change in name.

**Comment by cloud-fan** [comment]
It's simple to write to noop sink: `df.write.format("noop").save`, why do we need this extra public API?

**Comment by felixcheung** [comment]
> @rxin, this runs the query up to the point where `materialize` is called. The underlying RDD can then pick up from the last shuffle the next time it is used. This works better than caching in most cases when using dynamic allocation because executors are not sitting idle, but work can be resumed and shared across queries. We could rename the method if that would be more clear.
> 
> @srowen, I've seen this suggested on the dev list a few times and I think it is a good idea to add it. There is not guarantee that `count` does the same thing -- it could be optimized -- and it is a little tricky to get this to work with the dataset API. This version creates a new DataFrame from the underlying RDD so that the work is reused from the last shuffle, instead of allowing the planner to re-optimize with later changes (usually projections) and discard the intermediate result. We have found this really useful for better control over the planner, as well as to cache data using the shuffle system.

I have to agree with this - I've seen `count()` or `cache()` mis-used too many times and too many times people need to go back to clean up and remove all calls to `count()`. So much so I'm planning to write a optimizer rule to remove them. I'm only partly kidding.

Maybe this isn't the API for it, and that's ok, let's improve it then and make good suggestion to the community/contributor etc.

I'm not sure `df.write.format("noop").save` is a good suggestion to general spark user.

**Comment by rdblue** [comment]
I think this should be an action, not a sink. A no-op sink is just another way to misuse existing APIs for a different purpose. And worse, a noop sink doesn't actually accomplish the goal. This call returns a dataframe that will reuse the data stored in shuffle servers. A noop sink would not work for dataframes because you have to get Spark to re-use the same underlying RDD that has been run.

**Comment by cloud-fan** [comment]
I've spent more time understanding the use case, and think table cache should be a better choice here
1. disk vs memory: you can set the storage level to disk-only with more than one copy, which is more reliable than shuffle files.
2. shuffle service: it supports RDD blocks as well thanks to https://github.com/apache/spark/pull/24499

In addition, the table cache has more advantages:
1. It can work for any dataframes, even without shuffles
2. other queries can benefit from table cache implicitly

You do have a point that table cache is lazy, but we can add a flag to control it. e.g. `def cache(eager: Boolean = false)`

**Comment by rdblue** [comment]
@cloud-fan, I wasn't aware of #24499. Serving cache blocks from the shuffle service sounds like it could be a good alternative solution. I like that it would serve whole blocks instead of shards. We would definitely need a way to eagerly cache.

**Comment by dongjoon-hyun** [comment]
Hi, All.
Given the previous discussion and long inactivity, I'll close this PR.
@vgankidi . Feel free to reopen this if you have another opinion.

**Pull request closed without merging**